lib/kvdb/kvdb_ctxn.c

```diff
@@ -75,6 +75,20 @@ struct kvdb_ctxn_set_impl {
 
 /* clang-format on */
 
+// Would love to document context in which txn functions can be called.
+// Possibilities include:
+//   - hse init/fini, kvdb open/close, client txn operations, client cursor
+//     operations, reaper, and probably delayed work.
+//
+// With so many possibilities it is hard to reason about the correctness of this
+// code.
+//
+// How about a standard way of identifying these contexts that we can use
+// throughout our code base? At minimum there are two contexts (client and
+// internal) but a more specific classification might be helpful (eg: hse
+// init/free, kvdb open/close, REST, reaper, compaction, etc.).
```

Comment:
Ideally contracts like these should be enforced at an API level

Reply:
Interesting idea. Do you mean using thread local variables to identify each thread (eg: client_thread=true, cursor=true, reaper=false) with asserts to enforce it?

Reply:
I don't know if that is what I mean, but for instance, say you have an API that initializes your library, ala hse_init(). And no other functions of your API are valid unless the hse_init() has been called.

Then conceivably you can structure your API like so:

```c
hse_err_t hse_init(..., struct hse_handle **hdl);
hse_err_t hse_kvdb_open(struct hse_handle *hdl, ...);
```

Then the API essentially forces you to use the intended pattern. In this case, maybe your context is a static enum instance in the functions that call into kvdb_ctxn API for instance. By requiring this context, you force the developer to think about where and what the intended use-case is.